### DISCUSSION TITLE
Custom annotations

### DISCUSSION THREAD

**Discussion by janik-martin** [linkml/linkml#709] [Q&A]
Hi, I was trying to add my own custom annotations to some of my classes. Something like:
```
...
default_prefix: moucmn
...
classes:
  sample class:
    ...
    annotations:
      moucmn:my_annotation:
        tag: moucmn:my_annotation
        value: some value
```
but it skips the export with message:
`INFO:root:Ignoring namespace error: Unknown CURIE prefix: moucmn`

I found this same approach with [Biolink Model](https://biolink.github.io/biolink-model/) which is a huge schema (and source of some advanced features knowledge for me :) 
For example:
```
...
default_prefix: biolink
...
subclass of:
  is_a: related to at concept level
  ...
  annotations:
      biolink:canonical_predicate:
        tag: biolink:canonical_predicate
        value: true
      biolink:opposite_of:
        tag: biolink:opposite_of
        value: biolink:superclass of
```
I am not sure if I am doing something wrong, or this is not possible with linkml and biolink is an exception. Note that using 'standard' well known annotations works fine..

Thanks for any suggestions.

**Comment by cmungall** [comment]
Can I just clarify, you are seeing this behavior in gen-owl?

**Reply by janik-martin** [reply]
Ooh sorry, I forgot to mention that. Yes in gen-owl, in fact its the same method `add_metadata` in owlgen.py where we were adding the description export.. :) The main question here for me - is this suppose to work this way or no? Because if yes, I would be glad to look "deeper" into the issue..

**Reply by cmungall** [reply]
Yes, it would be good to include translation of annotations

Some caveats - the annotation system is intentionally quite loose - you can use unprefixed tags, and the mapping of these to URIs is unspecified. The values are modeled as simple strings, again without any typing.

Also, there is a syntactic trick that is optional but useful, you can say

```yaml
 annotations:
    tag: value
```

this maybe looks a little odd when prefixed tags are used

**Reply by janik-martin** [reply]
Sorry, not sure if I get the answer. But I did some digging in, and now I'm pretty sure using my own prefix in annotations won't work.
I am now mainly into owl gen and jsonld context gen, so I don't know the impact to other parts of export, but do you think it would be reasonable to export annotations without the prefix using the default_prefix (they are skipped in owl-gen right now)?

**Reply by cmungall** [reply]
yes, definitely. 

I think that makes sense, along with using untyped literals as values, as default.

however, we can also have a check that determines if the tag corresponds to an existing slot, and if so, use that interpretation (its slot_uri, and translate to literal or uri based on the range of that slot)

**Reply by janik-martin** [reply]
> however, we can also have a check that determines if the tag corresponds to an existing slot, and if so, use that interpretation (its slot_uri, and translate to literal or uri based on the range of that slot)

Actually I tried that yesterday and failed (but my knowing of linkml source code is very limited so far). In owlgen I tried using `_prop_uri` method to transform non-prefixed annotations into their URIs to add them into the RDF graph. 
The problem is with slots containing spaces in their names (which is a common practice in linkml), Annotation tags need to be URIorCURIEs which e.g. "sample annotation" tag is NOT (this fails with `ValueError: sample annotation is not a valid URI or CURIE`) while parsing schema. 

On the other hand, using its normalised underscore version e.g. `sample_annotation` then fails in `_prop_uri` method, where the dictionary schema.slots has the original name (with spaces) as key, so the lookup fails.

So it seems to me, that without a major rework, is is impossible to check the existence of the slot in the schema.. Please say I'm wrong and push me into the right direction :) 

Export the annotations unchecked, using literals as values seems possible though. This would be the same level of strictness as with the prefixed ones - using schema:whateverMadeOfAnnotation as a tag will export just fine..

**Reply by cmungall** [reply]
part of the problem here is that owlgen is an 'older style generator', schema introspection is easier with the newer schemaview-style generators

I know this info is not so useful to you in isolation, I will try and get back on this...
